browser/ai_chat/content_agent_tools_browsertest.cc

```diff
@@ -0,0 +1,423 @@
+    base::JSONWriter::Write(input, &json_string);
+    return json_string;
+  }
+
+  // Helper to execute a tool and wait for completion
+  Tool::ToolResult ExecuteToolAndWait(base::WeakPtr<Tool> tool,
+                                      const std::string& input_json,
+                                      bool verify_success = true) {
+    base::test::TestFuture<Tool::ToolResult> result_future;
+    tool->UseTool(input_json, result_future.GetCallback());
+    auto result = result_future.Take();
+    if (verify_success) {
+      EXPECT_THAT(result, ContentBlockText(
+                              testing::HasSubstr(kToolResultSuccessSubstring)));
+    }
+    return result;
+  }
+
+  // Helper to get real DOM node ID from a page element
+  int GetDOMNodeId(const std::string& selector) {
+    return content::GetDOMNodeId(*main_frame(), selector).value();
+  }
+
+  // Helper to get the web contents
+  content::WebContents* web_contents() {
+    base::test::TestFuture<tabs::TabHandle> tab_handle_future;
+    tool_provider_->GetOrCreateTabHandleForTask(
+        tab_handle_future.GetCallback());
+    tabs::TabHandle tab_handle = tab_handle_future.Take();
+    return tab_handle.Get()->GetContents();
+  }
+
+  // Helper to get the main frame
+  content::RenderFrameHost* main_frame() {
+    return web_contents()->GetPrimaryMainFrame();
+  }
+
+  // Helper to get the profile
+  Profile* GetProfile() { return agent_profile_; }
+
+  // Helper to navigate to Chromium test files
+  void NavigateToChromiumTestFile(const std::string& file_path) {
+    base::test::TestFuture<tabs::TabHandle> tab_handle_future;
+    tool_provider_->GetOrCreateTabHandleForTask(
+        tab_handle_future.GetCallback());
+    tabs::TabHandle tab_handle = tab_handle_future.Take();
+
+    GURL test_url = embedded_test_server()->GetURL(file_path);
+    ASSERT_TRUE(
+        content::NavigateToURL(tab_handle.Get()->GetContents(), test_url));
+  }
+
+  raw_ptr<Profile> agent_profile_;
+  std::unique_ptr<ContentAgentToolProvider> tool_provider_;
+  base::test::ScopedFeatureList scoped_feature_list_;
+};
+
+// Test click tool with Node ID targeting
+IN_PROC_BROWSER_TEST_F(ContentAgentToolsTest, ClickTool_NodeIdTarget) {
+  NavigateToChromiumTestFile("/actor/page_with_clickable_element.html");
+
+  auto click_tool = FindToolByName("click_element");
+  ASSERT_TRUE(click_tool);
+
+  // Reset click tracking
+  ASSERT_TRUE(content::ExecJs(web_contents(), "button_clicked = false"));
+  ASSERT_TRUE(content::ExecJs(web_contents(), "mouse_event_log = []"));
+
+  // Get real DOM node ID for the clickable button
+  int button_node_id = GetDOMNodeId("button#clickable");
+
+  auto target_dict = target_test_util::GetContentNodeTargetDict(
+      button_node_id,
+      *optimization_guide::DocumentIdentifierUserData::GetDocumentIdentifier(
+          web_contents()->GetPrimaryMainFrame()->GetGlobalFrameToken()));
+
+  base::Value::Dict input;
+  input.Set("target", std::move(target_dict));
```

Comment:
for consistency with the other tests should we be cloning here?

Reply:
touché! changed to clone 🤺